File: assets/timeserver.py
from datetime import datetime
import json

from sanic import Sanic, response



async def getTimeNow():
  res = datetime.now().astimezone().isoformat()
  print(f"getTimeNow(): {res}")
  return res

async def processToolCall(jso):
  print(jso)
  funcName = jso["name"]
  if funcName == "get_time_now":
    return await getTimeNow()
  else:
    return "unknown function: " & funcName

app = Sanic("timeServerApp")

@app.websocket("/ws")
async def ws(request, ws):
  async for msg in ws:
    jso = json.loads(msg)
    await ws.send(await processToolCall(jso))

@app.get("/")
async def functions(request):
  return response.json({"functions": [
    {
      "name": "get_time_now",
      "description": "get the current time in ISO format",
      "params": [],
      "returns": {
        "type": "string",
        "description": "current time in ISO format",
      }
    }
  ]})

if __name__ == "__main__":
  app.run(host="0.0.0.0", port=4234)

File: assets/timeserver.nim
import std/[
  asyncdispatch,
  asynchttpserver,
  json,
  logging,
  times,
]

import ws

import utils



tools:
  proc getTimeNow(): string =
    result = $(now())
    echo "getTimeNow(): ", result

proc processToolCall(jso: JsonNode): Future[string] {.async.} =
  logging.debug jso
  let funcName = jso["name"].getStr
  case funcName
  of "getTimeNow":
    return getTimeNow()
  else:
    return "unknown function: " & funcName

var toolDefs {.threadvar.}: JsonNode

proc cb(req: Request) {.async, gcsafe.} =
  if req.url.path == "/ws":
    try:
      var ws = await newWebSocket(req)
      while ws.readyState == Open:
        let jso = (await ws.receiveStrPacket()).parseJson
        let res = await processToolCall(jso)
        await ws.send($res)
    except:
      echo getCurrentExceptionMsg()
  elif req.url.path == "/":
    let headers = {
      "Content-type": "application/json; charset=utf-8"
    }.newHttpHeaders()
    await req.respond(Http200, $toolDefs, headers)
  else:
    await req.respond(Http404, "Not found")

when isMainModule:
  when not defined(release):
    addHandler(newConsoleLogger(levelThreshold = lvlDebug))

  toolDefs = toolFuncs
  var server = newAsyncHttpServer()
  const port = 2234
  logging.debug "Listening on port: ", port
  waitFor server.serve(Port(port), cb)
